util/jemalloc_nodump_allocator.cc

```diff
@@ -6,26 +6,98 @@
 #include "util/jemalloc_nodump_allocator.h"
 
 #include <string>
+#include <thread>
 
+#include "port/port.h"
 #include "util/string_util.h"
 
 namespace rocksdb {
+namespace jemalloc {
 
 #ifdef ROCKSDB_JEMALLOC_NODUMP_ALLOCATOR
 
 std::atomic<extent_alloc_t*> JemallocNodumpAllocator::original_alloc_{nullptr};
 
 JemallocNodumpAllocator::JemallocNodumpAllocator(
-    unsigned arena_index, int flags, std::unique_ptr<extent_hooks_t>&& hooks)
-    : arena_index_(arena_index), flags_(flags), hooks_(std::move(hooks)) {
-  assert(arena_index != 0);
+    PerCPUArena per_cpu_arena, bool enable_tcache, unsigned num_cpus,
+    std::vector<std::unique_ptr<extent_hooks_t>>&& arena_hooks,
+    std::vector<unsigned>&& arena_indices)
+    : per_cpu_arena_(per_cpu_arena),
+      enable_tcache_(enable_tcache),
+      num_cpus_(num_cpus),
+      arena_hooks_(std::move(arena_hooks)),
+      arena_indices_(std::move(arena_indices)),
+      tcache_(&JemallocNodumpAllocator::DestroyThreadSpecificCache) {
+  switch (per_cpu_arena) {
+    case PerCPUArena::kDisabled:
+      assert(arena_indices_.size() == 1);
+      break;
+    case PerCPUArena::kPerCPU:
+      assert(arena_indices_.size() == num_cpus_);
+      break;
+    case PerCPUArena::kPerPhysicalCPU:
+      assert(arena_indices_.size() == (num_cpus_ + 1) / 2);
+      break;
+  }
+  assert(arena_hooks.size() == arena_indices.size());
+  for (unsigned __attribute__((__unused__)) arena_index : arena_indices) {
+    assert(arena_index != 0);
+  }
+}
+
+int JemallocNodumpAllocator::GetThreadSpecificCache() {
+  if (!enable_tcache_) {
+    return MALLOCX_TCACHE_NONE;
+  }
+  unsigned* tcache_index = reinterpret_cast<unsigned*>(tcache_.Get());
+  if (tcache_index == nullptr) {
+    // Instantiate tcache.
+    tcache_index = new unsigned(0);
+    size_t tcache_index_size = sizeof(unsigned);
+    int ret =
+        mallctl("tcache.create", tcache_index, &tcache_index_size, nullptr, 0);
+    if (ret != 0) {
+      // No good way to expose the error. Silently disable tcache.
+      delete tcache_index;
+      return MALLOCX_TCACHE_NONE;
+    }
+    tcache_.Reset(static_cast<void*>(tcache_index));
+  }
+  return MALLOCX_TCACHE(*tcache_index);
 }
 
 void* JemallocNodumpAllocator::Allocate(size_t size) {
-  return mallocx(size, flags_);
+  // Obtain tcache.
+  int tcache_flag = GetThreadSpecificCache();
+
+  // Choose arena.
+  unsigned arena_id = 0;
+  if (per_cpu_arena_ != PerCPUArena::kDisabled) {
+    int cpu_id = port::PhysicalCoreID();
+    if (cpu_id < 0) {
+      // Failed to get CPU id.
+      arena_id = 0;
```

Comment:
The practice we are using in other places is that, we randomize which one to use if CPU ID is not available. See https://github.com/facebook/rocksdb/blob/189f0c27aaecdf17ae7fc1f826a423a28b77984f/util/core_local.h#L64-L75

Actually, why don't we directly use CoreLocalArray?

Reply:
`CoreLocalArray` does not allow me to:
1. Do physical-core-local (half the number of arenas in case of hyper-threading)
2. Initialize object for all cores (I want to be handwaving and create all arenas in the beginning, instead of doing lazy initialization).
I can update `CoreLocalArray` though.

Reply:
I must have missed conversation somewhere. What's the benefit of sharing arena for one hyper-threaded physical core?
Initializing objects for all cores can be done through CoreLocalArray, can we?

If you don't want to use the whole CoreLocalArray, at least reuse the exiting logic of CoreLocalArray<T>::AccessElementAndIndex(). They seem to be 90% the same as what you want to do here.